geoxarray/accessor.py

```diff
@@ -51,24 +52,80 @@ class _SharedGeoAccessor(object):
 
     def __init__(self, xarray_obj):
         self._obj = xarray_obj
-        # self._center = None
+        self._x_dim = None
+        self._y_dim = None
+        self._vertical_dim = None
+        self._time_dim = None
+        # map current dimension to geoxarray preferred dim name
+        self._dim_map = {}
+
         self._crs = None
-        self._is_gridded = True
-        self._cf_grid_mapping_parameters = None
+        self._is_gridded = None
+        self.set_dims()
 
-    @property
-    def cf_grid_mapping_parameters(self):
-        return self._cf_grid_mapping_parameters
+    def set_dims(self, x=None, y=None, vertical=None, time=None):
+        """Set preferred dimension names.
 
-    @cf_grid_mapping_parameters.setter
-    def cf_grid_mapping_parameters(self, value):
-        self._cf_grid_mapping_parameters = value
-        # reset the CRS object so it gets recalculated
-        self._crs = None
+        GeoXarray will use this information for future operations.
+        If any of the dimensions are not provided they will be found
+        by best guess.
 
-    def find_cf_grid_mapping(self):
-        """Search for CF standard grid mapping information."""
-        return self._cf_grid_mapping_parameters
+        Parameters
+        ----------
+
+        x : str or None
+            Name of the X dimension. This dimension usually exists with
+            a corresponding coordinate variable in meters for
+            gridded/projected data.
+        y : str or None
+            Name of the Y dimension. Similar to the X dimension but on the Y
+            axis.
+        vertical : str or None
+            Name of the vertical or Z dimension. This dimension usually exists
+            with a corresponding coordinate variable in meters for altitude
+            or pressure level (ex. hPa, millibar, etc).
+        time : str or None
+            Name of the time dimension. This dimension usually exists with a
+            corresponding coordinate variable with time objects.
+
+        """
+        obj = self._obj
+        dims = obj.dims
+        if x is None:
+            if 'x' in dims:
+                self._x_dim = 'x'
+        if y is None:
+            if 'y' in dims:
+                self._y_dim = 'y'
+        if vertical is None:
+            for z_dim in ('z', 'vertical', 'pressure_level'):
+                if z_dim in dims:
+                    self._vertical_dim = z_dim
+                    break
+        if time is None:
+            for t_dim in ('time', 't'):
+                if t_dim in dims:
+                    self._time_dim = t_dim
+                    break
+
+        self._dim_map[self._x_dim] = 'x'
```

Comment:
I am wondering if the dim map should be a property. The reasoning for this is that the dimensions could end up being `None` for all of them to start out with. In this case, they would all be overwritten in the dict. If it is a property, then you could check if they are none or not before accessing it and raise an exception if it is None or cache it if it is fine.

Reply:
Hm interesting. I'll try.

Reply:
So the only thing I don't like about this is recreating the dictionary every time. I could get around this by making a special dictionary subclass that handles these checks in its `__getitem__`. I'm wondering if I should use the `False/None` idea for the dimension handling too so a user could say `set_dims(x=False)` meaning "I know I have an 'x' dimension but it isn't the 'x' you think it is".

Reply:
You don't have to re-create the dictionary every time. If you made the dictionary a property that gets accessed, it would only be generated once (unless you call `set_dims` again, then it would set it to None and force it to be re-created) and cached on the call of the property. This way, it won't raise an exception when the accessor is created and allows the user to set the x/y/time/vertical dims and call the functions that use it later.